org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/GridView.java

```diff
@@ -613,6 +617,108 @@ public void clear()
 		this.getChildren().clear();
 	}
 
+	// name of attribute to identify menu item
+	public static final String ATTR_NAME_IDENTIFY_MENU_ITEM = "columnMenuIdentify";
+	// value of attribute to identify freeze menu item
+	public static final String ATTR_VALUE_MENU_ITEM_FROZEN = "Frozen";
+	public static final String MENU_ITEM_FROZEN_LABEL = "FrozenColumnContextMenuLabel";
+	// value of attribute to identify reset freeze menu item
+	public static final String ATTR_VALUE_MENU_ITEM_RESET_FROZEN = "Reset Frozen";
+	public static final String MENU_ITEM_RESET_FROZEN_LABEL = "ResetFrozenColumnContextMenuLabel";
+	// name of attribute to store reference to column doing right click contextMenuColumn
+	public static final String ATTR_NAME_CONTEXT_MENU_COLUMN = "contextMenuColumn";
+	// name of attribute to store reference to menu popup
+	public static final String ATTR_NAME_GRID_VIEW_CONTEXT_MENU_POPUP = "gridViewContextMenuPopup";
+	
+	/**
+	 * Event listener for context menu on grid header<br/>
+	 * Freeze the current column when click "frozen" column and reset to default column when click "reset frozen"
+	 */
+	public static final EventListener<? extends Event> listener = event -> {
+		Menuitem menuItem = (Menuitem)event.getTarget();
+		Menupopup menupopup = (Menupopup)menuItem.getParent();
+		
+		// get refers to the column set when handling the onOpen event of MenuPopop
+		org.zkoss.zul.Column contextMenuColumn = (org.zkoss.zul.Column)menupopup.getAttribute(GridView.ATTR_NAME_CONTEXT_MENU_COLUMN);
+		org.zkoss.zul.Grid grid = contextMenuColumn.getGrid();
+		
+		// calculate frozen up to click menu
+		int frozenColumns = -1;
+		Object valueIdentifyMenuItem = menuItem.getAttribute(ATTR_NAME_IDENTIFY_MENU_ITEM);
+		if (GridView.ATTR_VALUE_MENU_ITEM_FROZEN.equals(valueIdentifyMenuItem)) {// freeze column menu
+			frozenColumns = grid.getColumns().getChildren().indexOf(contextMenuColumn) + 1;
+		}else if (GridView.ATTR_VALUE_MENU_ITEM_RESET_FROZEN.equals(valueIdentifyMenuItem)) {// reset freeze menu
+			frozenColumns = 2;
+		}
+		
+		Frozen frozen = grid.getFrozen();
+		if (frozenColumns != -1 && frozen != null) {
+			frozen.setColumns(frozenColumns);
+		}
+	};
+	
+	/**
+	 * Assign context menu to grid column. Use one menu for all columns per each desktop <br/>
+	 * 1. lookup or create MenuPopup <br/>
+	 * 2. set context menu for compSetPopup<br/> 
+	 * @param compGetRoot component to lookup root component on desktop to lookup MenuPopup
+	 * @param compSetPopup column to set context menu
+	 */
+	public static void setColumnsMenuPopup (Component compGetRoot, XulElement compSetPopup) {
+		if (compGetRoot ==null || compSetPopup == null || compGetRoot.getRoot() == null)
+			return;
+		
+		Object objMenuPopup = compGetRoot.getRoot().getAttribute(GridView.ATTR_NAME_GRID_VIEW_CONTEXT_MENU_POPUP);
```

Comment:
why do we need to store it as a root attribute instead of just as a throw away component ? I means it is just a popup menu with 2 menu item.